src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,175 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkLast(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        for (T element : list) {
+            linkLast(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).value;
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> x = node(index);
+        T oldValue = x.value;
+        x.value = value;
+        return oldValue;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
+            Node<T> node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
-    public boolean remove(T t) {
+    public boolean remove(T value) {
+        if (value == null) {
```

Comment:
do we need this checking? Let's use only one for loop in this method

Reply:
done